src/validator/types/array.rs

```diff
@@ -0,0 +1,152 @@
+use balena_temen as temen;
+use serde_json::Value;
+
+use crate::{
+    schema::UniqueItems,
+    validator::{scope::ScopedSchema, state::ValidationState, Validator},
+};
+
+enum ValueWrapper<'a> {
+    Borrowed(&'a Value),
+    Owned(Value),
+}
+
+impl<'a> ValueWrapper<'a> {
+    fn value_ref(&'a self) -> &'a Value {
+        match self {
+            ValueWrapper::Borrowed(value) => value,
+            ValueWrapper::Owned(ref value) => value,
+        }
+    }
+}
+
+impl<'a> From<Value> for ValueWrapper<'a> {
+    fn from(x: Value) -> ValueWrapper<'a> {
+        ValueWrapper::Owned(x)
+    }
+}
+
+impl<'a> From<&'a Value> for ValueWrapper<'a> {
+    fn from(x: &'a Value) -> ValueWrapper<'a> {
+        ValueWrapper::Borrowed(x)
+    }
+}
+
+impl<'a, 'b> PartialEq<ValueWrapper<'a>> for ValueWrapper<'b> {
+    fn eq(&self, other: &ValueWrapper<'a>) -> bool {
+        crate::utils::value::eq(self.value_ref(), other.value_ref())
+    }
+}
+
+fn validate_items_uniqueness_by(scope: &ScopedSchema, items: &[Value], path: &str) -> ValidationState {
+    let engine = temen::Engine::default();
+    let mut ctx = temen::Context::default();
+    let position = temen::ast::Identifier::default();
+
+    let mut disctinct = vec![];
+    let mut state = ValidationState::new();
+
+    for item in items {
+        match engine.eval(path, &position, item, &mut ctx).map(ValueWrapper::from) {
+            Ok(wrapped) => {
+                if disctinct.contains(&wrapped) {
+                    state.push_error(scope.error("uniqueItems", path));
+                } else {
+                    disctinct.push(wrapped);
+                }
+            }
+            Err(_) => {
+                state.push_error(scope.error("uniqueItems", path));
+            }
+        }
+    }
+
+    state
+}
+
+fn validate_items_uniqueness(scope: &ScopedSchema, items: &[Value]) -> ValidationState {
+    let mut distinct = vec![];
+
+    let mut state = ValidationState::new();
+
+    for wrapped in items.iter().map(ValueWrapper::from) {
+        if distinct.contains(&wrapped) {
+            state.push_error(scope.error("uniqueItems", "expected unique items"));
+        } else {
+            distinct.push(wrapped);
+        }
+    }
+
+    state
+}
+
+fn validate_unique_items(scope: &ScopedSchema, values: &[Value]) -> ValidationState {
+    match scope.schema().unique_items() {
+        UniqueItems::Boolean(false) => ValidationState::new(),
+        UniqueItems::Boolean(true) => validate_items_uniqueness(scope, values),
+        UniqueItems::Paths(ref paths) => {
+            let mut state = ValidationState::new();
+
+            for path in paths {
+                state.extend(validate_items_uniqueness_by(scope, values, path));
+            }
+
+            state
+        }
+    }
+}
+
+pub fn validate_as_array(scope: &ScopedSchema, data: &Value) -> ValidationState {
```

Comment:
may be worth naming validation steps either in comments or by extracting functions per each step, as this can be harder to read right now because of the lenght of the function.

Reply:
Just pushed some comments there.